### PR TITLE
Open Library

### PR DESCRIPTION
**Pull request by franciscoabalan**
## Summary 
Connect with open library api and create a open library endpoint

## Trellocard
https://trello.com/c/0YlLMMCC/35-integraci%C3%B3n-api-externa

## Screenshot
Captura de Pantalla 2019-11-04 a la(s) 09 24 48

### CODE REVIEW

File: app/controllers/api/v1/books_controller.rb [hunk 1]
```diff
@@ -10,6 +10,10 @@ def index
       def show
         render json: Book.find(params[:id])
       end
+
+      def open_library
```
Comment (joredudiaz) [inline app/controllers/api/v1/books_controller.rb]: can we named in more descriptive way? like `search` or something like that

File: app/services/open_library_service.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    book = HTTParty.get(
+      "https://openlibrary.org/api/books?bibkeys=ISBN:#{isbn}&jscmd=data&format=json"
+    )
+    data = {}
+    data[:isbn] = isbn
+    data[:title] = book["ISBN:#{isbn}"]['title']
+    data[:subtitle] = book["ISBN:#{isbn}"]['subtitle']
+    data[:number_of_pages] = book["ISBN:#{isbn}"]['number_of_pages']
+    data[:authors] = book["ISBN:#{isbn}"]['authors']
+    data
```
Comment (joredudiaz) [inline app/services/open_library_service.rb]: what happen if book is not found? you must raise an exception with `book is not found`

File: app/services/open_library_service.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class OpenLibraryService
+  def self.execute(isbn)
+    book = HTTParty.get(
```
Comment (joredudiaz) [inline app/services/open_library_service.rb]: it is a good practice put the base url in a reusable var or you can `include` HTTParty gem on the class and use `base_uri` method, we use like that.

File: app/services/open_library_service.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+      "https://openlibrary.org/api/books?bibkeys=ISBN:#{isbn}&jscmd=data&format=json"
+    )
+    data = {}
+    data[:isbn] = isbn
```
Comment (joredudiaz) [inline app/services/open_library_service.rb]: we can do this method better, think if we need more fields from the service 🤔we can not add more and more lines to this method, we can find a better way 😄

**Review by joredudiaz — COMMENTED** [review]
it isn't wrong, but you can do the method more modular

File: app/controllers/api/v1/books_controller.rb (hunk 1 above)
Comment (franciscoabalan) [reply app/controllers/api/v1/books_controller.rb]: Done

File: app/services/open_library_service.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class OpenLibraryService
```
Comment (aleochoam) [inline app/services/open_library_service.rb:1]: As Jorge says, it's a good pratice that you include the HTTParty gem in the class, and call the methods from the class

File: app/services/open_library_service.rb (hunk 4 above)
Comment (aleochoam) [reply app/services/open_library_service.rb]: A way i think is to do this:

```
book["ISBN:#{isbn}"].slice('title', 'subtitle', 'number_of_pages', 'authors')
```

Checkout slice [here](https://apidock.com/rails/Hash/slice)

File: config/routes.rb [hunk 6]
```diff
@@ -3,6 +3,7 @@
   
   api_version(:module => "api/v1", path: { value: 'api/v1' }, defaults: { format: :json }) do
     resources :books,  only: %i[index show]
+    get '/open_library' => 'books#open_library', as: "open_library"
```
Comment (aleochoam) [inline config/routes.rb]: This route should be inside books,  that way is exposed like `api/v1/books/open_library`, don't you think?

File: app/services/open_library_service.rb (hunk 2 above)
Comment (franciscoabalan) [reply app/services/open_library_service.rb]: Done

File: app/services/open_library_service.rb (hunk 4 above)
Comment (franciscoabalan) [reply app/services/open_library_service.rb]: Done

File: config/routes.rb (hunk 6 above)
Comment (franciscoabalan) [reply config/routes.rb]: Very good comment.. thanks.

Done!

File: app/services/open_library_service.rb (hunk 3 above)
Comment (franciscoabalan) [reply app/services/open_library_service.rb]: Done

File: app/services/open_library_service.rb (hunk 5 above)
Comment (franciscoabalan) [reply app/services/open_library_service.rb:1]: Done

File: app/services/open_library_service.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  def self.execute(isbn)
+    include HTTParty
```
Comment (aleochoam) [inline app/services/open_library_service.rb]: Let's move this line to be  included at class level an not a  method  level

File: app/services/open_library_service.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    base_uri 'https://openlibrary.org/api/books'
```
Comment (aleochoam) [inline app/services/open_library_service.rb]: Same here

File: app/services/open_library_service.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    response = get('/', query: { bibkeys: "ISBN:#{isbn}", jscmd: 'data', format: 'json' })
```
Comment (aleochoam) [inline app/services/open_library_service.rb:6]: 👏

File: app/services/open_library_service.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+    if response["ISBN:#{isbn}"].present?
+      response["ISBN:#{isbn}"].slice('title', 'subtitle', 'number_of_pages', 'authors')
+    else
+      Exception.new('Book not Found')
```
Comment (aleochoam) [inline app/services/open_library_service.rb]: Are you catching this exception in the controller?

File: app/services/open_library_service.rb (hunk 7 above)
Comment (franciscoabalan) [reply app/services/open_library_service.rb]: Done

File: app/services/open_library_service.rb (hunk 8 above)
Comment (franciscoabalan) [reply app/services/open_library_service.rb]: Done

File: app/services/open_library_service.rb (hunk 10 above)
Comment (franciscoabalan) [reply app/services/open_library_service.rb]: I made a refactor and now is catching the record not found rescue in application controller

File: app/services/open_library_service.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
+class OpenLibraryService
+  include HTTParty
+  base_uri 'https://openlibrary.org/api/books'
+
+  def self.execute(isbn)
+    response = get('/', query: { bibkeys: "ISBN:#{isbn}", jscmd: 'data', format: 'json' })
+    raise ActiveRecord::RecordNotFound, 'Book not found' if response.parsed_response.empty?
```
Comment (aleochoam) [inline app/services/open_library_service.rb]: Mm i don't know if RecordNotFound is a good exception, this has nothing to do with the db, RuntimeError could be a good one

**Review by aleochoam — APPROVED** [review]

**Review by joredudiaz — APPROVED** [review]
👏 excellent!!! 😄

**Pull request merged**